lxd/profiles.go

```diff
@@ -43,11 +43,70 @@ var profilesCmd = APIEndpoint{
 var profileCmd = APIEndpoint{
 	Path: "profiles/{name}",
 
-	Delete: APIEndpointAction{Handler: profileDelete, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanDelete, "name")},
-	Get:    APIEndpointAction{Handler: profileGet, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanView, "name")},
-	Patch:  APIEndpointAction{Handler: profilePatch, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
-	Post:   APIEndpointAction{Handler: profilePost, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
-	Put:    APIEndpointAction{Handler: profilePut, AccessHandler: allowPermission(entity.TypeProfile, auth.EntitlementCanEdit, "name")},
+	Delete: APIEndpointAction{Handler: profileDelete, AccessHandler: profileAccessHandler(auth.EntitlementCanDelete)},
+	Get:    APIEndpointAction{Handler: profileGet, AccessHandler: profileAccessHandler(auth.EntitlementCanView)},
+	Patch:  APIEndpointAction{Handler: profilePatch, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+	Post:   APIEndpointAction{Handler: profilePost, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+	Put:    APIEndpointAction{Handler: profilePut, AccessHandler: profileAccessHandler(auth.EntitlementCanEdit)},
+}
+
+// ctxProfileDetails should be used only for getting/setting profileDetails in the request context.
+const ctxProfileDetails request.CtxKey = "profile-details"
+
+// profileDetails contains fields that are determined prior to the access check. This is set in the request context when
+// addProfileDetailsToRequestContext is called.
+type profileDetails struct {
+	profileName      string
+	effectiveProject api.Project
+}
+
+// addProfileDetailsToRequestContext sets request.CtxEffectiveProjectName (string) and ctxProfileDetails (profileDetails)
+// in the request context.
+func addProfileDetailsToRequestContext(d *Daemon, r *http.Request) error {
```

Comment:
Is there a reason you're requiring *Daemon here, rather than *state.State as an argument? It doesn't look like you're using d for any other reason that to access d.State(), which could be passed in from the caller, which better indicates the requirements for this function.

Reply:
Fixed. I should have been using state, however the distinction between `*Deamon` and `*state.State` isn't clear to me in LXD. If the state is a snapshot of the daemon state at the time the request is received, I'd say this should be evaluated at the top-level mux handler passed into both the access handler and main handler.